File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/fs/VfsRootRadio.java
package app.zxtune.fs;

import android.content.Context;
import android.net.Uri;

import androidx.annotation.Nullable;

import java.io.IOException;
import java.util.ArrayList;
import java.util.concurrent.Callable;

import javax.annotation.CheckForNull;

import app.zxtune.Log;
import app.zxtune.R;

@Icon(R.drawable.ic_browser_vfs_radio)
final class VfsRootRadio extends StubObject implements VfsRoot {

  private static final String TAG = VfsRootRadio.class.getName();

  private final Uri URI = Uri.fromParts("radio", "", "");

  private final Context ctx;
  @CheckForNull
  private ArrayList<VfsDir> children;

  VfsRootRadio(Context ctx) {
    this.ctx = ctx;
  }

  @Nullable
  @Override
  public VfsObject resolve(Uri uri) {
    if (URI.equals(uri)) {
      return this;
    } else {
      return null;
    }
  }

  @Override
  public void enumerate(Visitor visitor) {
    for (VfsDir child : getChildren()) {
      visitor.onDir(child);
    }
  }

  @Override
  public Uri getUri() {
    return URI;
  }

  @Override
  public String getName() {
    return ctx.getString(R.string.vfs_radio_root_name);
  }

  @Override
  public String getDescription() {
    return ctx.getString(R.string.vfs_radio_root_description);
  }

  @Nullable
  @Override
  public VfsObject getParent() {
    return null;
  }

  private ArrayList<VfsDir> getChildren() {
    if (children == null) {
      children = new ArrayList<>();
      maybeAdd(children, EntryModarchive::new);
      maybeAdd(children, EntryVgmrips::new);
    }
    return children;
  }

  private static void maybeAdd(ArrayList<VfsDir> list, Callable<VfsDir> create) {
    try {
      list.add(create.call());
    } catch (Exception e) {
      Log.w(TAG, e, "Failed to add radio source");
    }
  }

  private static class BaseEntry extends StubObject implements VfsDir {

    private final VfsDir delegate;

    BaseEntry(String uri) throws IOException {
      this.delegate = (VfsDir) Vfs.resolve(Uri.parse(uri));
    }

    @Override
    public Uri getUri() {
      return delegate.getUri();
    }

    @Override
    public String getName() {
      return delegate.getName();
    }

    @Override
    public String getDescription() {
      return delegate.getDescription();
    }

    @Override
    public VfsObject getParent() {
      // not required really
      return delegate.getParent();
    }

    @Override
    public Object getExtension(String id) {
      return delegate.getExtension(id);
    }

    @Override
    public void enumerate(Visitor visitor) throws IOException {
      delegate.enumerate(visitor);
    }
  }

  // TODO: get rid of wrappers after browser rework
  @Icon(R.drawable.ic_browser_vfs_modarchive)
  private class EntryModarchive extends BaseEntry {

    EntryModarchive() throws IOException {
      super("modarchive:/Random");
    }

    @Override
    public String getDescription() {
      return ctx.getString(R.string.vfs_modarchive_root_name);
    }
  }

  @Icon(R.drawable.ic_browser_vfs_vgmrips)
  private class EntryVgmrips extends BaseEntry {

    EntryVgmrips() throws IOException {
      super("vgmrips:/Random");
    }

    @Override
    public String getDescription() {
      return ctx.getString(R.string.vfs_vgmrips_root_name);
    }
  }
}

File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/fs/VfsRootRadio.kt
package app.zxtune.fs

import android.content.Context
import android.net.Uri
import androidx.annotation.DrawableRes
import androidx.annotation.StringRes
import androidx.core.net.toUri
import app.zxtune.Logger
import app.zxtune.R

internal class VfsRootRadio(private val ctx: Context) : StubObject(), VfsRoot {
    override val uri: Uri = Uri.fromParts("radio", "", "")
    private val children by lazy {
        ArrayList<VfsDir>().apply {
            maybeAdd(
                this,
                "modarchive:/Random",
                R.string.vfs_modarchive_root_name,
                R.drawable.ic_browser_vfs_modarchive
            )
            maybeAdd(
                this,
                "vgmrips:/Random",
                R.string.vfs_vgmrips_root_name,
                R.drawable.ic_browser_vfs_vgmrips
            )
            maybeAdd(
                this,
                "khinsider:/Random",
                R.string.vfs_khinsider_root_name,
                R.drawable.ic_browser_vfs_khinsider
            )
        }
    }

    override val name: String
        get() = ctx.getString(R.string.vfs_radio_root_name)
    override val description: String
        get() = ctx.getString(R.string.vfs_radio_root_description)
    override val parent: VfsObject?
        get() = null

    override fun getExtension(id: String) = when (id) {
        VfsExtensions.ICON -> R.drawable.ic_browser_vfs_radio
        else -> super.getExtension(id)
    }

    override fun resolve(uri: Uri): VfsObject? = this.takeIf { it.uri == uri }

    override fun enumerate(visitor: VfsDir.Visitor) = children.forEach(visitor::onDir)

    private fun maybeAdd(
        list: ArrayList<VfsDir>, uri: String, @StringRes description: Int, @DrawableRes icon: Int
    ) = runCatching {
        list.add(BaseEntry(uri, description, icon))
    }.onFailure { e ->
        LOG.w(e) { "Failed to add radio source" }
    }

    private inner class BaseEntry(
        private val delegate: VfsDir,
        @StringRes private val descr: Int,
        @DrawableRes private val icon: Int
    ) : StubObject(), VfsDir by delegate {

        constructor(
            uri: String, @StringRes descr: Int, @DrawableRes icon: Int
        ) : this(Vfs.resolve(uri.toUri()) as VfsDir, descr, icon)

        override val description
            get() = ctx.getString(descr)

        override fun getExtension(id: String) = when (id) {
            VfsExtensions.ICON -> icon
            else -> delegate.getExtension(id)
        }
    }

    companion object {
        private val LOG = Logger(VfsRootRadio::class.java.name)
    }
}
